crates/iceberg/src/spec/values.rs

```diff
@@ -0,0 +1,941 @@
+                        (object.remove("keys"), object.remove("values"))
+                    {
+                        Ok(Literal::Map(BTreeMap::from_iter(
+                            keys.into_iter()
+                                .zip(values.into_iter())
+                                .map(|(key, value)| {
+                                    Ok((
+                                        Literal::try_from_json(key, &map.key_field.field_type)?,
+                                        Some(Literal::try_from_json(
+                                            value,
+                                            &map.value_field.field_type,
+                                        )?),
+                                    ))
+                                })
+                                .collect::<Result<Vec<_>, Error>>()?
+                                .into_iter(),
+                        )))
+                    } else {
+                        Err(Error::new(
+                            crate::ErrorKind::DataInvalid,
+                            "The json value for a list type must be an array.",
+                        ))
+                    }
+                } else {
+                    Err(Error::new(
+                        crate::ErrorKind::DataInvalid,
+                        "The json value for a list type must be an array.",
+                    ))
+                }
+            }
+        }
+    }
+
+    /// Get datatype of value
+    pub fn datatype(&self) -> Type {
+        match self {
+            Literal::Primitive(prim) => match prim {
+                PrimitiveLiteral::Boolean(_) => Type::Primitive(PrimitiveType::Boolean),
+                PrimitiveLiteral::Int(_) => Type::Primitive(PrimitiveType::Int),
+                PrimitiveLiteral::Long(_) => Type::Primitive(PrimitiveType::Long),
+                PrimitiveLiteral::Float(_) => Type::Primitive(PrimitiveType::Float),
+                PrimitiveLiteral::Double(_) => Type::Primitive(PrimitiveType::Double),
+                PrimitiveLiteral::Date(_) => Type::Primitive(PrimitiveType::Date),
+                PrimitiveLiteral::Time(_) => Type::Primitive(PrimitiveType::Time),
+                PrimitiveLiteral::Timestamp(_) => Type::Primitive(PrimitiveType::Timestamp),
+                PrimitiveLiteral::TimestampTZ(_) => Type::Primitive(PrimitiveType::Timestamptz),
+                PrimitiveLiteral::Fixed(vec) => {
+                    Type::Primitive(PrimitiveType::Fixed(vec.len() as u64))
+                }
+                PrimitiveLiteral::Binary(_) => Type::Primitive(PrimitiveType::Binary),
+                PrimitiveLiteral::String(_) => Type::Primitive(PrimitiveType::String),
+                PrimitiveLiteral::UUID(_) => Type::Primitive(PrimitiveType::Uuid),
+                PrimitiveLiteral::Decimal(dec) => Type::Primitive(PrimitiveType::Decimal {
+                    precision: 38,
```

Comment:
This isn't always 38. It actually depends on the number of bytes that are read. But we can break that out in a separate chore

Reply:
This is how we do it in Python:

```python
MAX_PRECISION = tuple(math.floor(math.log10(math.fabs(math.pow(2, 8 * pos - 1) - 1))) for pos in range(24))
REQUIRED_LENGTH = tuple(next(pos for pos in range(24) if p <= MAX_PRECISION[pos]) for p in range(40))


def decimal_required_bytes(precision: int) -> int:
    """Compute the number of bytes required to store a precision.

    Args:
        precision: The number of digits to store.

    Returns:
        The number of bytes required to store a decimal with a certain precision.
    """
    if precision <= 0 or precision >= 40:
        raise ValueError(f"Unsupported precision, outside of (0, 40]: {precision}")

    return REQUIRED_LENGTH[precision]
```

And how we do it in Java: https://github.com/apache/iceberg/blob/f5f543a54ff7460648bb864f4f06a29eb28938b9/api/src/main/java/org/apache/iceberg/types/TypeUtil.java#L679-L715